### ISSUE TITLE
Discrepancies in ERR extension when processing bias frames with CALACS

### ISSUE THREAD

**Issue by nmiles2718** [spacetelescope/hstcal#277]
While verifying the changes made to `ACSREJ` in #2, I noticed that for bias calibration frames (exptime=0) there is an existing bug in the current release of `hstcal` concerning the computation of values in the ERR extensions. There are currently two ways to batch process and combine files with the `calacs` codebase; using an [association file](http://www.stsci.edu/hst/acs/documents/handbooks/currentDHB/acs_Ch22.html#149623), or using the python wrapper [acstools](http://acstools.readthedocs.io/en/latest/) for `calacs` to process the files and then using python wrapper for [ACSREJ.](http://acstools.readthedocs.io/en/latest/acsrej.html) 

If the raw bias frames are combined using an association with only `DQICORR`,`BLEVCORR`,`CRCORR` set to `PERFORM` the error arrays are correctly computed. However, if the bias frames are combined by first processing each image individually with only `DQICORR` and `BLEVCORR` set to `PERFORM`, and then using the `acstools` wrapper for acsrej there is an additional Poisson-like term that is included in the error which should not be present for images with EXPTIME=0. The wrapper only serves to facilitate user input and in the end is a direct call to the `acsrej.e` executable (defined at [L130](https://github.com/spacetelescope/acstools/blob/3025736de199951fcc346ba95385309dac9ba077/acstools/acsrej.py#L130), called at [L187](https://github.com/spacetelescope/acstools/blob/3025736de199951fcc346ba95385309dac9ba077/acstools/acsrej.py#L187)), which is the same executable that `calacs` uses.

![error_v_signal_hstcal_bias](https://user-images.githubusercontent.com/20194427/32788567-92c8e97c-c927-11e7-867e-7a60edaaf87e.png)

In the figure above the two populations of blue dots (one for the read noise of each amplifier used to readout chip 2) are the correct ones, whereas the red ones have an additional poisson-like term.

I double checked to ensure that the problem only exists for the bias frames by performing the same analysis on a set of dark frames with an exptime of 1000.5s. The two images produced are essentially identical with only 0.0025% of the pixels differing by more than 0.1e in a 1000's exposure where the typical pixel has a value less than 140 e-. The two plots below show the pixel values in the final images produced by each process; the first is the full spectrum, the second is zoomed in to show the ~97% of pixels with values less than 140e- which is the the hot pixel threshold in 1000.5s dark frames.
![error_v_signal_hstcal_dark](https://user-images.githubusercontent.com/20194427/32796946-2d4ca842-c93e-11e7-92c8-a7a257eb4fb5.png)

![error_v_signal_hstcal_dark_zoomed](https://user-images.githubusercontent.com/20194427/32796848-f23c7c6e-c93d-11e7-9501-59893018c651.png)

**Comment by pllim** [comment]
This might or might not be a waste of time: What happens when you only use `acsccd.e` on those BIAS frames to make `_blv_tmp.fits` and then use `acsrej.e` on those instead of FLT?

(Although if Michele finds the source of the bug before you can get to this, feel free to ignore.)

**Comment by mdlpstsci** [comment]
There is a “behind the scenes with no message to the trailer file” check on the bias processing in CALACS.  It checks whether the ACS HRC/WFC bias data being processed is pre- or post-SM4 (2009-01-01).  This check sets a flag to tell the cosmic ray rejection loop how to process.  The dataset jd4z* which I copied from you is from July 2017 which is clearly post-SM4. 
 
In contrast the ACSREJ standalone routine has a command line switch, -newbias, which sets the same flag.  If the data is post-SM4, the –newbias option should be set on the ACSREJ command line.  I am guessing that you did not set the –newbias flag, and therefore, incorrect (or perhaps I should say different) processing was applied to your data. 
 
I processed the jd4z* data with CALACS and with ACSREJ standalone, both with the –newbias option present and not present when using ACSREJ.  The CALACS results seem to match (used fitsdiff which found some low-level differences) to the ACSREJ –newbias results, but I really need your verification on this.
 
The ACSREJ standalone executable does not offer any explanation of the command line switches in the help syntax which I can fix to clarify.   Also, CALACS does not issue an appropriate message to the trailer file when it modifies a default flag setting based upon "behind the scenes machinations".
 
I will wait to hear from you regarding verification of this issue.

**Comment by pllim** [comment]
Wow, good investigation, @mdlpstsci ! The `newbias` stuff was added way back in 641309a75c4ab453913c9ace8 by @stsci-hack according to `git blame`. I was still in INS at that point and not involved in pipeline, hence I can't explain this but maybe Warren still remembers why.

If I have to make an educated guess, it would be due to the bias gradient introduced by the new electronics installed during SM4.

**Comment by mdlpstsci** [comment]
Here is some commentary I got from Warren after some forensic research on his part:

The change was implemented on 3 Feb 2010 as documented on https://trac.stsci.edu/ssb/stsci_python/changeset/8712.

Specifically, all this reports is:
Updates were made to ACSREJ to add a new parameter (newbias) to support 
processing of post-SM4 bias images. Changes were also made throughout 
CALACS to automatically determine this parameter value and pass it to 
ACSREJ, which used this value to change how the error arrays are 
computed for CRJ products created with post-SM4 bias images.

Unfortunately, I could not find any emails discussing the need or reasoning behind this specific change, but it appears to come as a result of the changes to ACS readout modes or data as a result of SM4.  The ACS Instrument Handbook may have some reference to this change, but that is a real long-shot.

That is the best I can come up with,
Warren
